### ISSUE TITLE
Security token

### ISSUE THREAD

**Issue by zhanglaoye** [clousale/amazon-sp-api-php#8]
When I access getordes, the security token is missing and an error is reported as unauthorized
![1606615793(1)](https://user-images.githubusercontent.com/47973176/100530279-0d1e3c80-322b-11eb-938a-1b5cbf3e73b5.png)

**Comment by lyubo-slavilov** [comment]
Same here.
As far as I understood from what I read in the net (I have no previouse experience with MWS or AWS) there is an additional step to `assumRole` before sending the request. I am not completly sure how to execute this in the context of this library.
I was not able make any progress beyond this point. Anybody?

**Comment by clousale** [comment]
We're currently checking with Amazon what could cause this. I'll keep you updated.

**Comment by lyubo-slavilov** [comment]
I was able to crack it by using the STS client from aws-php-sdk to create a session for the role ARN. The session token I've injected as `amz-security-token` header into the SignatureSellingPartner::calculateSignature() method;

This was just a POC as it requires all the calls to SignatureSellingPartner::calculateSignature() to be changed in order to accept one more argument.
Same for the all API calls as they need to provide this `amz-security-token`

I can provide a simple pseudocode for the process you need assistance

**Comment by clousale** [comment]
Hey @lyubo-slavilov
Thanks for the status update. We're doing something similar currently, but imho that's a really inefficient way, because we have to `composer require` a large library (aws sdk) just for a minor signature calculation. As soon as we have a solution for that without having to require the aws sdk, we will update this package. Also, as said, we're currently talking to Amazon to get a solution.
In the meantime, to help others as well, some example code is surely appreciated @lyubo-slavilov

**Comment by lyubo-slavilov** [comment]
@clousale Hi there, I was able to replicate the `assumeRole` action of the STS service.
Here are the crucial parts:

## Reverse engineering
I have used `Aws\Sts\StsClient` to examine the request sent to the sts endpoint.
Here is an example debug output:
```http
POST / HTTP/1.1
Host: sts.amazonaws.com
Content-Type: application/x-www-form-urlencoded
aws-sdk-invocation-id: e35c7eb7ab3cfc4f6e1f8c1863088e42
aws-sdk-retry: 0/0
X-Amz-Date: 20201202T144016Z
Authorization: AWS4-HMAC-SHA256 Credential=<YOUR-ARN-API-KEY_HERE>/20201202/us-east-1/sts/aws4_request, SignedHeaders=host;x-amz-date, Signature=d68da58b9667bc5d...ad06a5cfbf33e04496a31a0
User-Agent: aws-sdk-php/3.164.1 GuzzleHttp/6.5.5 curl/7.70.0 PHP/7.4.9
Content-Length: 123

Action=AssumeRole&RoleArn=<YOUR-IAM-ROLE-ARN>&RoleSessionName=<ARBITRARY-NAME>&Version=2011-06-15
```

## Replicate the results
Now, It seem as a simple request, but is is signed one. I have adopted your `SignatureSellingPartner` class to be more flexible  in the sign process (mostly to accept less or more headers) and I have prepared custom request to STS:

```php
  $requestOptions = [
      'headers' => [
        'accept' => 'application/json'
      ],
      'form_params' => [
        'Action' => 'AssumeRole',
        'DurationSeconds' => 3600, // TODO: load this from config
        'RoleArn' => '<YOUR-IAM-ROLE-ARN>' ,
        'RoleSessionName' => 'session1',
        'Version' => '2011-06-15',
      ]
    ];

    $host = 'sts.amazonaws.com';
    $uri = '/';

    $signer = new MySigner();
    $requestOptions = $signer->sign($requestOptions, [
      'service' => 'sts',
      'access_key' => '<YOUR-ARN-API-KEY>',
      'secret_key' => '<YOUR-ARN-APY-SECRET>',
      'region' => 'us-east-1', //Should be this hardcoded?
      'host' => $host,
      'uri' => $uri,
      'payload' => GuzzleHttp\Psr7\build_query($requestOptions['form_params']),
      'method' => 'POST',
    ]);

    $client = new GuzzleHttp\Client([
      'base_uri' => 'https://' . $host
    ]);

    $response = $client->post($uri, $requestOptions);
    $json = json_decode($response->getBody(), true);

    print_r($json['AssumeRoleResponse']['AssumeRoleResult']['Credentials']);
```

This will output something like:
```
Array(
  [AccessKeyId] => <STS-ACCESS-KEY>,
  [SecretAccessKey] => <STS-SECRET-KEY>,
  [SessionToken] => <STS-SESSION-TOKEN>
)
```

## Ready to go
So, now we are ready to make requests to SP API. 
- First we need to use `<STS-ACCESS-KEY>` and `<STS-SECRET-KEY>`  (instead of `<YOUR-ARN-API-KEY>` and `<YOUR-ARN-APY-SECRET>`) when signing the requests
- Second we need to pass the `<STS-SESSION-TOKEN>` as `aws-security-token` header in the request, and it needs to be in the `SignedHeaders` set.

## Results
I was able successfully to prepare custom requests to OrdersV0 API and I got expected results. Unfortunately all of this is crafts and arts and I was mostly working with row clients, requests and responses. It is not easy to retrofit these changes to your API client. 

## Side notes
1. I discovered that your signer does not handle properly the request payload when the content-type is `application/x-www-form-urlencoded`. It produces excessive quotes around the query string. as example if your have 
``` php
$requestOptions = [
  'form_params' => ["type" => "electron", "charge" => "negative"] 
]
```
The sign process eventually will use `"type=electron&charge=negative"` for the payload hashing instead of `type=electron&charge=negative`
2. STS sessions are short lived but they are not one-time sessions. So a mechanism for  storing STS session keys will be needed
3. If you want to try `Aws\Sts\StsClient` I recommend to check how the php sdk handles its configuration. It was deep rabbit hole. Eventually I have found you need to place your `<YOUR-ARN-API-KEY>` and `<YOUR-ARN-APY-SECRET>` in `~/.aws/credentials` ini file in order initial STS requests to be signed properly.

Cheers.

**Comment by clousale** [comment]
Hey @lyubo-slavilov thanks for your amazing post.
We will try and make this package convenient for all users based on your findings.

**Comment by ykw090721** [comment]
@clousale Waiting for your update to amazon-sp-api-php

**Comment by lzz29** [comment]
> @clousale Hi there,
> 
> **TL;DR**
> I was able to replicate the `assumeRole` action of the STS service and i was able to make requests to SP API without using any code from aws-php-sdk
> 
> Here are the crucial parts:
> 
> ## Reverse engineering
> I have used `Aws\Sts\StsClient` to examine the request sent to the sts endpoint.
> Here is an example debug output:
> 
> ```httpspec
> POST / HTTP/1.1
> Host: sts.amazonaws.com
> Content-Type: application/x-www-form-urlencoded
> aws-sdk-invocation-id: e35c7eb7ab3cfc4f6e1f8c1863088e42
> aws-sdk-retry: 0/0
> X-Amz-Date: 20201202T144016Z
> Authorization: AWS4-HMAC-SHA256 Credential=<YOUR-ARN-API-KEY_HERE>/20201202/us-east-1/sts/aws4_request, SignedHeaders=host;x-amz-date, Signature=d68da58b9667bc5d...ad06a5cfbf33e04496a31a0
> User-Agent: aws-sdk-php/3.164.1 GuzzleHttp/6.5.5 curl/7.70.0 PHP/7.4.9
> Content-Length: 123
> 
> Action=AssumeRole&RoleArn=<YOUR-IAM-ROLE-ARN>&RoleSessionName=<ARBITRARY-NAME>&Version=2011-06-15
> ```
> 
> ## Replicate the results
> Now, It seem as a simple request, but is is signed one. I have adopted your `SignatureSellingPartner` class to be more flexible in the sign process (mostly to accept less or more headers) and I have prepared custom request to STS:
> 
> ```
>   $requestOptions = [
>       'headers' => [
>         'accept' => 'application/json'
>       ],
>       'form_params' => [
>         // Please note parameters are in lexical ascending order. This is crucial for computing proper signature
>         'Action' => 'AssumeRole',
>         'DurationSeconds' => 3600, // TODO: load this from config
>         'RoleArn' => '<YOUR-IAM-ROLE-ARN>' ,
>         'RoleSessionName' => 'session1',
>         'Version' => '2011-06-15',
>       ]
>     ];
> 
>     $host = 'sts.amazonaws.com';
>     $uri = '/';
> 
>     $signer = new MySigner();
>     $requestOptions = $signer->sign($requestOptions, [
>       'service' => 'sts',
>       'access_key' => '<YOUR-ARN-API-KEY>',
>       'secret_key' => '<YOUR-ARN-APY-SECRET>',
>       'region' => 'us-east-1', //Should be this hardcoded?
>       'host' => $host,
>       'uri' => $uri,
>       'payload' => GuzzleHttp\Psr7\build_query($requestOptions['form_params']),
>       'method' => 'POST',
>     ]);
> 
>     $client = new GuzzleHttp\Client([
>       'base_uri' => 'https://' . $host
>     ]);
> 
>     $response = $client->post($uri, $requestOptions);
>     $json = json_decode($response->getBody(), true);
> 
>     print_r($json['AssumeRoleResponse']['AssumeRoleResult']['Credentials']);
> ```
> 
> This will output something like:
> 
> ```
> Array(
>   [AccessKeyId] => <STS-ACCESS-KEY>,
>   [SecretAccessKey] => <STS-SECRET-KEY>,
>   [SessionToken] => <STS-SESSION-TOKEN>
> )
> ```
> 
> ## Ready to go
> So, now we are ready to make requests to SP API.
> 
> * First we need to use `<STS-ACCESS-KEY>` and `<STS-SECRET-KEY>`  (instead of `<YOUR-ARN-API-KEY>` and `<YOUR-ARN-APY-SECRET>`) when signing the requests
> * Second we need to pass the `<STS-SESSION-TOKEN>` as `x-aws-security-token` header in the request, and it needs to be in the `SignedHeaders` set.
> * Third, we continue as before and provide the clients with LWA access token, etc...
> 
> ## Results
> I was able successfully to prepare custom requests to OrdersV0 API and I got expected results. Unfortunately all of this is crafts and arts and I was mostly working with raw clients, requests and responses. It is not easy to retrofit these changes to your API client.
> 
> ## Side notes
> 1. I discovered that your signer does not handle properly the request payload when the content-type is `application/x-www-form-urlencoded`. It produces excessive quotes around the query string. as example if your have
> 
> ```
> $requestOptions = [
>   'form_params' => ["type" => "electron", "charge" => "negative"] 
> ]
> ```
> 
> The sign process eventually will use `"type=electron&charge=negative"` for the payload hashing instead of `type=electron&charge=negative`
> 2. STS sessions are short lived but they are not one-time sessions. So a mechanism for storing STS session keys will be needed
> 3. If you want to try `Aws\Sts\StsClient` I recommend to check how the php sdk handles its configuration. It was deep rabbit hole. Eventually I have found you need to place your `<YOUR-ARN-API-KEY>` and `<YOUR-ARN-APY-SECRET>` in `~/.aws/credentials` ini file in order initial STS requests to be signed properly.
> 
> Cheers.



> @clousale Hi there,
> 
> **TL;DR**
> I was able to replicate the `assumeRole` action of the STS service and i was able to make requests to SP API without using any code from aws-php-sdk
> 
> Here are the crucial parts:
> 
> ## Reverse engineering
> I have used `Aws\Sts\StsClient` to examine the request sent to the sts endpoint.
> Here is an example debug output:
> 
> ```httpspec
> POST / HTTP/1.1
> Host: sts.amazonaws.com
> Content-Type: application/x-www-form-urlencoded
> aws-sdk-invocation-id: e35c7eb7ab3cfc4f6e1f8c1863088e42
> aws-sdk-retry: 0/0
> X-Amz-Date: 20201202T144016Z
> Authorization: AWS4-HMAC-SHA256 Credential=<YOUR-ARN-API-KEY_HERE>/20201202/us-east-1/sts/aws4_request, SignedHeaders=host;x-amz-date, Signature=d68da58b9667bc5d...ad06a5cfbf33e04496a31a0
> User-Agent: aws-sdk-php/3.164.1 GuzzleHttp/6.5.5 curl/7.70.0 PHP/7.4.9
> Content-Length: 123
> 
> Action=AssumeRole&RoleArn=<YOUR-IAM-ROLE-ARN>&RoleSessionName=<ARBITRARY-NAME>&Version=2011-06-15
> ```
> 
> ## Replicate the results
> Now, It seem as a simple request, but is is signed one. I have adopted your `SignatureSellingPartner` class to be more flexible in the sign process (mostly to accept less or more headers) and I have prepared custom request to STS:
> 
> ```
>   $requestOptions = [
>       'headers' => [
>         'accept' => 'application/json'
>       ],
>       'form_params' => [
>         // Please note parameters are in lexical ascending order. This is crucial for computing proper signature
>         'Action' => 'AssumeRole',
>         'DurationSeconds' => 3600, // TODO: load this from config
>         'RoleArn' => '<YOUR-IAM-ROLE-ARN>' ,
>         'RoleSessionName' => 'session1',
>         'Version' => '2011-06-15',
>       ]
>     ];
> 
>     $host = 'sts.amazonaws.com';
>     $uri = '/';
> 
>     $signer = new MySigner();
>     $requestOptions = $signer->sign($requestOptions, [
>       'service' => 'sts',
>       'access_key' => '<YOUR-ARN-API-KEY>',
>       'secret_key' => '<YOUR-ARN-APY-SECRET>',
>       'region' => 'us-east-1', //Should be this hardcoded?
>       'host' => $host,
>       'uri' => $uri,
>       'payload' => GuzzleHttp\Psr7\build_query($requestOptions['form_params']),
>       'method' => 'POST',
>     ]);
> 
>     $client = new GuzzleHttp\Client([
>       'base_uri' => 'https://' . $host
>     ]);
> 
>     $response = $client->post($uri, $requestOptions);
>     $json = json_decode($response->getBody(), true);
> 
>     print_r($json['AssumeRoleResponse']['AssumeRoleResult']['Credentials']);
> ```
> 
> This will output something like:
> 
> ```
> Array(
>   [AccessKeyId] => <STS-ACCESS-KEY>,
>   [SecretAccessKey] => <STS-SECRET-KEY>,
>   [SessionToken] => <STS-SESSION-TOKEN>
> )
> ```
> 
> ## Ready to go
> So, now we are ready to make requests to SP API.
> 
> * First we need to use `<STS-ACCESS-KEY>` and `<STS-SECRET-KEY>`  (instead of `<YOUR-ARN-API-KEY>` and `<YOUR-ARN-APY-SECRET>`) when signing the requests
> * Second we need to pass the `<STS-SESSION-TOKEN>` as `x-aws-security-token` header in the request, and it needs to be in the `SignedHeaders` set.
> * Third, we continue as before and provide the clients with LWA access token, etc...
> 
> ## Results
> I was able successfully to prepare custom requests to OrdersV0 API and I got expected results. Unfortunately all of this is crafts and arts and I was mostly working with raw clients, requests and responses. It is not easy to retrofit these changes to your API client.
> 
> ## Side notes
> 1. I discovered that your signer does not handle properly the request payload when the content-type is `application/x-www-form-urlencoded`. It produces excessive quotes around the query string. as example if your have
> 
> ```
> $requestOptions = [
>   'form_params' => ["type" => "electron", "charge" => "negative"] 
> ]
> ```
> 
> The sign process eventually will use `"type=electron&charge=negative"` for the payload hashing instead of `type=electron&charge=negative`
> 2. STS sessions are short lived but they are not one-time sessions. So a mechanism for storing STS session keys will be needed
> 3. If you want to try `Aws\Sts\StsClient` I recommend to check how the php sdk handles its configuration. It was deep rabbit hole. Eventually I have found you need to place your `<YOUR-ARN-API-KEY>` and `<YOUR-ARN-APY-SECRET>` in `~/.aws/credentials` ini file in order initial STS requests to be signed properly.
> 
> Cheers.
Hello
Can you provide the code for MySinger Class?

**Comment by clousale** [comment]
@lyubo-slavilov would you mind submitting a PR for this?

**Comment by clousale** [comment]
Quick fix is using the `role` as `IAM ARN` in your application (as registered in seller central) as described in this comment:
https://github.com/amzn/selling-partner-api-docs/issues/24#issuecomment-744016448

We're still working on the `user` as `IAM ARN` which has the downside to send requests to `assumeRole` additionally.

**Comment by lyubo-slavilov** [comment]
@clousale, @lzz29  It was easier to write it from scratch without using SwaggerHub's code generator tool. It is is simply not designed to work with such complicated authorization mechanizms as Amazon's.

I've tried to came up with clever way to fix this situation here, but eventualy I gave up.
You can check this one: https://github.com/double-break/spapi-php

**Comment by kingsun-he** [comment]
> @clousale, @lzz29 It was easier to write it from scratch without using SwaggerHub's code generator tool. It is is simply not designed to work with such complicated authorization mechanizms as Amazon's.
> 
> I've tried to came up with clever way to fix this situation here, but eventualy I gave up.
> You can check my take here: https://github.com/double-break/spapi-php

Why did you give up? Does your fix work? @lyubo-slavilov

**Comment by lyubo-slavilov** [comment]
Yes, my lib is independent from SwaggerGen.
I've created my own generator.

**Comment by kingsun-he** [comment]
When I run index.php. This error happen:





Warning: curl_setopt_array(): cannot represent a stream of type Output as a STDIO FILE* in C:\xampp\htdocs\spapiphp\vendor\guzzlehttp\guzzle\src\Handler\CurlFactory.php on line 59

Warning: curl_setopt_array(): cannot represent a stream of type Output as a STDIO FILE* in C:\xampp\htdocs\spapiphp\vendor\guzzlehttp\guzzle\src\Handler\CurlFactory.php on line 59

Warning: curl_setopt_array(): cannot represent a stream of type Output as a STDIO FILE* in C:\xampp\htdocs\spapiphp\vendor\guzzlehttp\guzzle\src\Handler\CurlFactory.php on line 59

Fatal error: Uncaught GuzzleHttp\Exception\RequestException: cURL error 0: The cURL request was retried 3 times and did not succeed. The most likely reason for the failure is that cURL was unable to rewind the body of the request and subsequent retries resulted in the same error. Turn on the debug option to see what went wrong. See https://bugs.php.net/bug.php?id=47204 for more information. (see https://curl.haxx.se/libcurl/c/libcurl-errors.html) in C:\xampp\htdocs\spapiphp\vendor\guzzlehttp\guzzle\src\Handler\CurlFactory.php:201 Stack trace: #0 C:\xampp\htdocs\spapiphp\vendor\guzzlehttp\guzzle\src\Handler\CurlFactory.php(537): GuzzleHttp\Handler\CurlFactory::createRejection(Object(GuzzleHttp\Handler\EasyHandle), Array) #1 C:\xampp\htdocs\spapiphp\vendor\guzzlehttp\guzzle\src\Handler\CurlFactory.php(152): GuzzleHttp\Handler\CurlFactory::retryFailedRewind(Object(GuzzleHttp\Handler\CurlHandler), Object(GuzzleHttp\Handler\EasyHandle), Array) #2 C:\xampp\htdocs\spapiphp\vendor\guzzlehttp\guzzle\src\Handler\CurlFactory.php(105 in C:\xampp\htdocs\spapiphp\vendor\guzzlehttp\guzzle\src\Handler\CurlFactory.php on line 201







Any idea?







At 2021-01-22 00:37:07, "Lyubomir Slavilov" <notifications@github.com> wrote:

Yes, my lib is independent from SwaggerGen.
I've created my own generator.

**Comment by lyubo-slavilov** [comment]
You are in the wrong repository.
If you have issues with double-break/spapi-php, report it there.

Kind Regards,
*Lyubomir Slavilov*

**Comment by clousale** [comment]
We also decided to decouple from the Swagger Codegen, but took the base of the one-time generated code.
Feel free to upgrade to 2.0!

**Comment by kingsun-he** [comment]
> We also decided to decouple from the Swagger Codegen, but took the base of the one-time generated code.
> Feel free to upgrade to 2.0!

Hi clousale,  It seems that other API don't update except CatalogApi. @clousale
